keras_hub/src/models/sam/sam_layers.py

```diff
@@ -0,0 +1,396 @@
+        activation: str, optional. The activation for the mlp block's output
+            layer. Defaults to "relu".
+    """
+
+    def __init__(
+        self,
+        num_heads,
+        key_dim,
+        intermediate_dim,
+        skip_first_layer_pe,
+        attention_downsample_rate=2,
+        activation="relu",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.num_heads = num_heads
+        self.key_dim = key_dim
+        self.intermediate_dim = intermediate_dim
+        self.skip_first_layer_pe = skip_first_layer_pe
+        self.attention_downsample_rate = attention_downsample_rate
+        self.activation = activation
+
+        self.self_attention = MultiHeadAttentionWithDownsampling(
+            num_heads=num_heads, key_dim=key_dim, dtype=self.dtype_policy
+        )
+        self.layer_norm1 = keras.layers.LayerNormalization(
+            epsilon=1e-5, dtype=self.dtype_policy
+        )
+        self.cross_attention_token_to_image = (
+            MultiHeadAttentionWithDownsampling(
+                num_heads=num_heads,
+                key_dim=key_dim,
+                downsample_rate=attention_downsample_rate,
+                dtype=self.dtype_policy,
+            )
+        )
+        self.layer_norm2 = keras.layers.LayerNormalization(
+            epsilon=1e-5, dtype=self.dtype_policy
+        )
+
+        self.mlp_block = MLP(
+            intermediate_dim,
+            key_dim * num_heads,
+            num_layers=2,
+            activation=activation,
+            dtype=self.dtype_policy,
+        )
+
+        self.layer_norm3 = keras.layers.LayerNormalization(
+            epsilon=1e-5, dtype=self.dtype_policy
+        )
+        self.cross_attention_image_to_token = (
+            MultiHeadAttentionWithDownsampling(
+                num_heads=num_heads,
+                key_dim=key_dim,
+                downsample_rate=attention_downsample_rate,
+                dtype=self.dtype_policy,
+            )
+        )
+        self.layer_norm4 = keras.layers.LayerNormalization(
+            epsilon=1e-5, dtype=self.dtype_policy
+        )
+
+    def build(self, input_shape=None):
+        self.self_attention.build()
+        self.layer_norm1.build([None, None, self.num_heads * self.key_dim])
+        self.cross_attention_token_to_image.build()
+        self.layer_norm2.build([None, None, self.num_heads * self.key_dim])
+        self.mlp_block.build([None, None, self.num_heads * self.key_dim])
+        self.layer_norm3.build([None, None, self.num_heads * self.key_dim])
+        self.cross_attention_image_to_token.build()
+        self.layer_norm4.build([None, None, self.num_heads * self.key_dim])
+        self.built = True
+
+    def call(self, queries, keys, query_pe, key_pe):
+        if self.skip_first_layer_pe:
+            queries = self.self_attention(
+                query=queries, value=queries, key=queries
+            )
+        else:
+            queries_with_pe = queries + query_pe
```

Comment:
What does `pe` stand for here?

Reply:
positional embeddings. I updated the names wherever we have `_pe`